gandalf.js

```diff
@@ -1,5 +1,5 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?